src/profile-logic/call-node-info.js

```diff
@@ -274,51 +270,298 @@ export class CallNodeInfoImpl implements CallNodeInfo {
+  // the k'th prefix node of suffixOrderedCallNodes[i] is stored at deepNodes[i - suffixOrderIndexRangeStart[x]].
+  deepNodesForSuffixOrderIndexRange: Array<Uint32Array | null>, // IndexIntoInvertedNonRootCallNodeTable -> (Uint32Array | null)
+
+  depth: number[], // IndexIntoInvertedNonRootCallNodeTable -> number
+  length: number,
+|};
+
+// Compute the InvertedRootCallNodeTable.
+// We compute this information upfront for all roots. The root count is fixed -
+// the number of roots is the same as the number of functions in the funcTable.
+function _createInvertedRootCallNodeTable(
+  callNodeTable: CallNodeTable,
+  rootSuffixOrderIndexRangeEndCol: Uint32Array,
+  suffixOrderedCallNodes: Uint32Array,
+  defaultCategory: IndexIntoCategoryList
+): InvertedRootCallNodeTable {
+  const funcCount = rootSuffixOrderIndexRangeEndCol.length;
+  const category = new Int32Array(funcCount);
+  const subcategory = new Int32Array(funcCount);
+  const innerWindowID = new Float64Array(funcCount);
+  const sourceFramesInlinedIntoSymbol = new Array(funcCount);
+  let previousRootSuffixOrderIndexRangeEnd = 0;
+  for (let funcIndex = 0; funcIndex < funcCount; funcIndex++) {
+    const callNodesuffixOrderIndexRangeStart =
+      previousRootSuffixOrderIndexRangeEnd;
+    const callNodesuffixOrderIndexRangeEnd =
+      rootSuffixOrderIndexRangeEndCol[funcIndex];
+    previousRootSuffixOrderIndexRangeEnd = callNodesuffixOrderIndexRangeEnd;
+    if (
+      callNodesuffixOrderIndexRangeStart === callNodesuffixOrderIndexRangeEnd
+    ) {
+      sourceFramesInlinedIntoSymbol[funcIndex] = null;
+      // Leave the remaining columns at zero for this root.
+      continue;
+    }
+
+    // Fill the remaining fields with the conflict-resolved versions of the values
+    // in the non-inverted call node table.
+    const firstNonInvertedCallNodeIndex =
+      suffixOrderedCallNodes[callNodesuffixOrderIndexRangeStart];
+    let resolvedCategory =
+      callNodeTable.category[firstNonInvertedCallNodeIndex];
+    let resolvedSubcategory =
+      callNodeTable.subcategory[firstNonInvertedCallNodeIndex];
+    const resolvedInnerWindowID =
+      callNodeTable.innerWindowID[firstNonInvertedCallNodeIndex];
+    let resolvedSourceFramesInlinedIntoSymbol =
+      callNodeTable.sourceFramesInlinedIntoSymbol[
+        firstNonInvertedCallNodeIndex
+      ];
+
+    // Resolve conflicts in the same way as for the non-inverted call node table.
+    for (
+      let orderingIndex = callNodesuffixOrderIndexRangeStart + 1;
+      orderingIndex < callNodesuffixOrderIndexRangeEnd;
+      orderingIndex++
+    ) {
+      const currentNonInvertedCallNodeIndex =
+        suffixOrderedCallNodes[orderingIndex];
+      // Resolve category conflicts, by resetting a conflicting subcategory or
+      // category to the default category.
+      if (
+        resolvedCategory !==
+        callNodeTable.category[currentNonInvertedCallNodeIndex]
+      ) {
```

Comment:
Heh, I'm not sure if that's a fix or a regression. The new approach is less "precise", because there are now two stages at which category conflicts are resolved. With the old approach, we resolved conflicts only when turning the inverted stack table into the inverted call node table. Now we resolve conflicts both when turning the non-inverted stack table into the non-inverted call node table, and then again when creating an inverted call node based on the non-inverted call node table. I think in practice the difference won't matter, but there is a difference; the new approach may have more gray call nodes.

Example:

```
Non-inverted stack table:

3  0 A:1 blue
2  1   B:1 green
1  1     C:1 yellow
1  0   B:2 red
1  1     D:1 yellow

Inverted call node table:

1  B green -> gray
0    A blue
1  C yellow
0    B green -> gray
0      A blue
1  D yellow
0    B red -> gray
0      A blue
```